README.md
# Pydantic Examples

[![FullStackWithLawrence](https://a11ybadges.com/badge?text=FullStackWithLawrence&badgeColor=orange&logo=youtube&logoColor=282828)](https://www.youtube.com/@FullStackWithLawrence)<br>
[![Python](https://a11ybadges.com/badge?logo=python)](https://www.python.org/)
[![Pydantic](https://a11ybadges.com/badge?text=Pydantic&badgeColor=e92063)](https://www.langchain.com/)
[![JSON](https://a11ybadges.com/badge?logo=json)](https://www.json.org/json-en.html)<br>
[![Release Notes](https://img.shields.io/github/release/FullStackWithLawrence/pydantic-example)](https://github.com/FullStackWithLawrence/pydantic-example/releases)
![GHA pushMain Status](https://img.shields.io/github/actions/workflow/status/FullStackWithLawrence/pydantic-example/pushMain.yml?branch=main)
[![AGPL License](https://img.shields.io/github/license/overhangio/tutor.svg?style=flat-square)](https://www.gnu.org/licenses/agpl-3.0.en.html)
[![hack.d Lawrence McDaniel](https://img.shields.io/badge/hack.d-Lawrence%20McDaniel-orange.svg)](https://lawrencemcdaniel.com)

[Pydantic](https://docs.pydantic.dev/latest/) is the most widely used data validation library for Python. This repo demonstrates three popular use cases for Pydantic:

1. **Validation**. Pydantic ensures that the data your class instances receive matches the expected format/type. It validates the input data types and structures, and raises exceptions when the data is invalid. See the [example Python class](./grader/grader.py) in this repo.

2. **Data Parsing and Serialization**. See the [JSON validator](./grader/langchain.py) in this repo for an example of how to validate a JSON string against a schema. Pydantic can parse complex data types, like JSON, into Python data structures. It can also serialize Python objects back into JSON.

3. **Exception Handling**. See the [custom exceptions](./grader/exceptions.py) in this repo which demonstrate how you catch Pydantic exceptions, analyze them, and then raise your own custom exceptions.

## Installation

```console
git clone https://github.com/FullStackWithLawrence/pydantic-example.git
cd pydantic-example
make init
source venv/bin/activate
```

## Usage

```console
# command-line help
python3 -m grader.batch -h

# example usage
python3 -m grader.batch 'path/to/homework/json/files/'
```

### About This Example

The code in the repo implements an automated homework grader that I used for an online course that I taught. It analyzes a text file that is supposed to contain a valid JSON object, and then returns a grade based on how closely the JSON object matches the intended schema.

### Rubric

Rubric values are expressed as floats between 0 and 1.00, and can be overridden with environment variables.

```console
AG_INCORRECT_RESPONSE_TYPE_PENALTY_PCT=0.10
AG_INCORRECT_RESPONSE_VALUE_PENALTY_PCT=0.15
AG_RESPONSE_FAILED_PENALTY_PCT=0.20
AG_INVALID_RESPONSE_STRUCTURE_PENALTY_PCT=0.30
AG_INVALID_JSON_RESPONSE_PENALTY_PCT=0.50
```

### Expected output

```console
% done! Graded 10 assignments. Output files are in path/to/homework/json/files/out
```

<!-- prettier-ignore -->
```json
{
  "grade": 100,
  "message": "Great job!",
  "message_type": "Success"
}
```

<!-- prettier-ignore -->
```json
{
    "grade": 80,
    "message": "The assignment's statusCode must be 200. received: 403",
    "message_type": "ResponseFailedError"
}
```

<!-- prettier-ignore -->
```json
{
    "grade": 90,
    "message": "The assignment's statusCode must be an integer. received: <class 'str'>",
    "message_type": "IncorrectResponseTypeError"
}
```

<!-- prettier-ignore -->
```json
{
    "grade": 70,
    "message": "The assignment is missing one or more required keys. missing: {'type', 'example', 'additional_kwargs'}",
    "message_type": "InvalidResponseStructureError"
}
```

<!-- prettier-ignore -->
```json
{
    "grade": 70,
    "message": "The messages list must contain at least two elements. messages: [{'content': \"Oh, how delightful. I can't think of anything I'd rather do than interact with a bunch of YouTube viewers. Just kidding, I'd rather be doing literally anything else. But go ahead, introduce me to your lovely audience. I'm sure they'll be absolutely thrilled to meet me.\", 'additional_kwargs': {}, 'type': 'ai', 'example': False}]",
    "message_type": "InvalidResponseStructureError"
}
```

<!-- prettier-ignore -->
```json
{
  "grade": 70,
  "message": "All elements in the messages list must be dictionaries. messages: ['bad', 'data']",
  "message_type": "InvalidResponseStructureError"
}
```

<!-- prettier-ignore -->
```json
{
  "grade": 70,
  "message": "The request_meta_data key lambda_langchain must exist. request_meta_data: {}",
  "message_type": "InvalidResponseStructureError"
}
```

## Contributing

This project uses a mostly automated pull request and unit testing process. See the resources in .github for additional details. You additionally should ensure that pre-commit is installed and working correctly on your dev machine by running the following command from the root of the repo.

```console
pre-commit run --all-files
```

Pull requests should pass these tests before being submitted:

```console
make test
```

### Developer setup

```console
git clone https://github.com/lpm0073/automatic-grader.git
cd automatic-grader
make init
make activate
```


CHANGELOG.md
# [1.6.0](https://github.com/FullStackWithLawrence/pydantic-example/compare/v1.5.0...v1.6.0) (2023-12-04)


### Features

* force a new release ([e1caa63](https://github.com/FullStackWithLawrence/pydantic-example/commit/e1caa634d6ee318afa606e0d6dcc19cae72d6688))

# [1.5.0](https://github.com/lpm0073/automatic-grader/compare/v1.4.0...v1.5.0) (2023-11-29)

### Features

- all grading logic moved to pydantic ([b327ca7](https://github.com/lpm0073/automatic-grader/commit/b327ca70c8c5b8d22719b9848b2112ae1cdf8423))

# [1.4.0](https://github.com/lpm0073/automatic-grader/compare/v1.3.1...v1.4.0) (2023-11-29)

### Features

- add Pydantic classes to validate assignment structure and business rules ([33de272](https://github.com/lpm0073/automatic-grader/commit/33de2727b96cf2b60efbdfdefae72991ad7bb3d4))

## [1.3.1](https://github.com/lpm0073/automatic-grader/compare/v1.3.0...v1.3.1) (2023-11-28)

### Bug Fixes

- [@field](https://github.com/field)\_validator can only be used on class methods ([3f7824e](https://github.com/lpm0073/automatic-grader/commit/3f7824ee99d41f639b61cc569e377b411473162f))

# [1.3.0](https://github.com/lpm0073/automatic-grader/compare/v1.2.0...v1.3.0) (2023-11-28)

### Features

- add configurable rubric ([fdc3da8](https://github.com/lpm0073/automatic-grader/commit/fdc3da80d0dafd36c78f840a393a00bdca5b04b3))

# [1.2.0](https://github.com/lpm0073/automatic-grader/compare/v1.1.6...v1.2.0) (2023-11-28)

### Bug Fixes

- wrap **version**.py value in double quotes so that it matches black rules ([88f7214](https://github.com/lpm0073/automatic-grader/commit/88f72149c5f97d9fe3ad6129f96580ae1db9f995))

### Features

- add a batch assignment grader. ([e7df088](https://github.com/lpm0073/automatic-grader/commit/e7df088c842e37f622aa4670f483448cdeb3495b))

## [1.1.6](https://github.com/lpm0073/automatic-grader/compare/v1.1.5...v1.1.6) (2023-11-28)

### Bug Fixes

- ensure that **version**.py construction matches that of dev ([af9aa7f](https://github.com/lpm0073/automatic-grader/commit/af9aa7faf5e6782cae0ec914dc820b2d0779fb11))

## [1.1.5](https://github.com/lpm0073/automatic-grader/compare/v1.1.4...v1.1.5) (2023-11-28)

### Bug Fixes

- remove pre-commit tests from ci ([22ffa1b](https://github.com/lpm0073/automatic-grader/commit/22ffa1b1924a8da8dd725dff04496b0ede7e1f44))

## [1.1.4](https://github.com/lpm0073/automatic-grader/compare/v1.1.3...v1.1.4) (2023-11-28)

### Bug Fixes

- add dummy variables for aws credentials ([1ebf9ce](https://github.com/lpm0073/automatic-grader/commit/1ebf9cecd3577053356271d0289a335f1e9dd39c))
- create the file, append to the file ([7f2b4d3](https://github.com/lpm0073/automatic-grader/commit/7f2b4d3e581903df65727a97a2d5b67d296a3268))

## [1.1.3](https://github.com/lpm0073/automatic-grader/compare/v1.1.2...v1.1.3) (2023-11-28)

### Bug Fixes

- add shell ([3604be4](https://github.com/lpm0073/automatic-grader/commit/3604be4f70eb709993fe4f91d8edda87642ad005))
- git pull before push ([fad074a](https://github.com/lpm0073/automatic-grader/commit/fad074a9b799efdf3869e98baceb724c54ce83c0))
- run prettier with npx ([a46356e](https://github.com/lpm0073/automatic-grader/commit/a46356ef9bd3378114a1036dcc8c6eff8e016597))

## [1.1.2](https://github.com/lpm0073/automatic-grader/compare/v1.1.1...v1.1.2) (2023-11-28)

### Bug Fixes

- add shell ([0f03b26](https://github.com/lpm0073/automatic-grader/commit/0f03b2655cbf4ffd32d65cf1ac45773bd05d2cb1))

## [1.1.1](https://github.com/lpm0073/automatic-grader/compare/v1.1.0...v1.1.1) (2023-11-28)

### Bug Fixes

- add stronger type checking, and existence of dict keys, and verbose error output ([1ee48c5](https://github.com/lpm0073/automatic-grader/commit/1ee48c511cd1030bbf7201f7d58f28e25fea7463))
- remove terraform and reactjs tests ([a512a3a](https://github.com/lpm0073/automatic-grader/commit/a512a3a64d6f7fb9adee8464de10d35bc7ea6ead))
- switch from pre-commit-ci to pre-commit command line in bash ([3699ef2](https://github.com/lpm0073/automatic-grader/commit/3699ef27d0481c7284a6e9c1bf9715347258d7d4))

# [1.1.0](https://github.com/lpm0073/automatic-grader/compare/v1.0.1...v1.1.0) (2023-11-27)

### Features

- add unit testing ([4de144e](https://github.com/lpm0073/automatic-grader/commit/4de144e9116ccd097d3d501745ca1ddd72d104e2))
- test incorrect response file type ([666e686](https://github.com/lpm0073/automatic-grader/commit/666e6862fad8e1cbe2e4b0c6dd0eeb9493d89a79))

## [1.0.1](https://github.com/lpm0073/automatic-grader/compare/v1.0.0...v1.0.1) (2023-11-27)

### Bug Fixes

- make npm private so that it doesn't try to publish to npm ([a446fbe](https://github.com/lpm0073/automatic-grader/commit/a446fbe29de9f4281620078e1fd6da2e4e0bb17c))
- make npm private so that it doesn't try to publish to npm ([0ac8782](https://github.com/lpm0073/automatic-grader/commit/0ac8782cf6f06fee9662a08710d9f5fc3b8b2fb1))

# 1.0.0 (2023-11-27)

### Features

- add packaging functionality ([bb6856c](https://github.com/lpm0073/automatic-grader/commit/bb6856c99159cd24d818175a7acb0709fb5237c1))
- setup repo ([85ec05c](https://github.com/lpm0073/automatic-grader/commit/85ec05c972bf45d2cc0a1998dea2dde0704c7411))


grader/config.py
# -*- coding: utf-8 -*-
"""Setup for automated_grader package."""

import os

from dotenv import find_dotenv, load_dotenv
from pydantic import BaseModel, Field


# for local development and unit testing
dotenv_path = find_dotenv()
if os.path.exists(dotenv_path):
    load_dotenv(dotenv_path=dotenv_path, verbose=True)


class ConfigurationError(Exception):
    """Exception raised for errors in the configuration.

    Attributes:
        message -- explanation of the error
    """

    def __init__(self, message):
        self.message = message
        super().__init__(self.message)


# pylint: disable=too-few-public-methods
class _AGRubric(BaseModel):
    """Private class to strongly type and bound package configuration params."""

    INCORRECT_RESPONSE_TYPE_PENALTY_PCT: float = Field(0.10, ge=0.00, le=1.00)
    INCORRECT_RESPONSE_VALUE_PENALTY_PCT: float = Field(0.15, ge=0.00, le=1.00)
    RESPONSE_FAILED_PENALTY_PCT: float = Field(0.20, ge=0.00, le=1.00)
    INVALID_RESPONSE_STRUCTURE_PENALTY_PCT: float = Field(0.30, ge=0.00, le=1.00)
    INVALID_JSON_RESPONSE_PENALTY_PCT: float = Field(0.50, ge=0.00, le=1.00)

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        try:
            self.INCORRECT_RESPONSE_TYPE_PENALTY_PCT = float(
                os.getenv("AG_INCORRECT_RESPONSE_TYPE_PENALTY_PCT", "0.10")
            )
            self.INCORRECT_RESPONSE_VALUE_PENALTY_PCT = float(
                os.getenv("AG_INCORRECT_RESPONSE_VALUE_PENALTY_PCT", "0.15")
            )
            self.RESPONSE_FAILED_PENALTY_PCT = float(os.getenv("AG_RESPONSE_FAILED_PENALTY_PCT", "0.20"))
            self.INVALID_RESPONSE_STRUCTURE_PENALTY_PCT = float(
                os.getenv("AG_INVALID_RESPONSE_STRUCTURE_PENALTY_PCT", "0.30")
            )
            self.INVALID_JSON_RESPONSE_PENALTY_PCT = float(os.getenv("AG_INVALID_JSON_RESPONSE_PENALTY_PCT", "0.50"))
        except ValueError as err:
            raise ConfigurationError(
                f"Invalid configuration. Note that penalties should be between 0.00 and 1.00: {err}"
            ) from err


ag_rubric = _AGRubric()


class AGRubric:
    """Public class to access package configuration params."""

    INCORRECT_RESPONSE_TYPE_PENALTY_PCT = ag_rubric.INCORRECT_RESPONSE_TYPE_PENALTY_PCT
    INCORRECT_RESPONSE_VALUE_PENALTY_PCT = ag_rubric.INCORRECT_RESPONSE_VALUE_PENALTY_PCT
    RESPONSE_FAILED_PENALTY_PCT = ag_rubric.RESPONSE_FAILED_PENALTY_PCT
    INVALID_RESPONSE_STRUCTURE_PENALTY_PCT = ag_rubric.INVALID_RESPONSE_STRUCTURE_PENALTY_PCT
    INVALID_JSON_RESPONSE_PENALTY_PCT = ag_rubric.INVALID_JSON_RESPONSE_PENALTY_PCT
